atlasdb-cassandra/src/test/java/com/palantir/atlasdb/cassandra/ReloadingCloseableContainerTest.java

```diff
@@ -0,0 +1,328 @@
+            awaitOrFail(runWithFutureRunsFirstBarrier);
+            container.close();
+        });
+        Awaitility.await().during(Duration.ofSeconds(1)).until(() -> !closeFuture.isDone());
+        verifyNoInteractions(mockCloseable);
+
+        manuallyFinishRunWith.await();
+        Awaitility.await().atMost(Duration.ofSeconds(1)).until(closeFuture::isDone);
+    }
+
+    @Test
+    public void refreshCloseBlockedDuringRunWithInvocation()
+            throws IOException, BrokenBarrierException, InterruptedException {
+        ReloadingCloseableContainer<Closeable> container = getContainer();
+
+        CyclicBarrier runWithFutureRunsFirstBarrier = new CyclicBarrier(2);
+        CyclicBarrier manuallyFinishRunWith = new CyclicBarrier(2);
+
+        executor.submit(() -> container.executeWithResource(closeable -> {
+            awaitOrFail(runWithFutureRunsFirstBarrier);
+            awaitOrFail(manuallyFinishRunWith);
+            verifyNoInteractions(closeable);
+        }));
+
+        Future<?> refreshableFuture = executor.submit(() -> {
+            awaitOrFail(runWithFutureRunsFirstBarrier);
+            settableRefreshable.update(UPDATED_VALUE);
+        });
+
+        Awaitility.await().during(Duration.ofSeconds(1)).until(() -> !refreshableFuture.isDone());
+        verifyNoInteractions(mockCloseable);
+
+        manuallyFinishRunWith.await();
+
+        Awaitility.await().atMost(Duration.ofSeconds(1)).until(refreshableFuture::isDone);
+        verify(mockCloseable).close();
+    }
+
+    @Test
+    public void refreshUpdatesResource() {
+        ReloadingCloseableContainer<Closeable> container = getContainer();
+
+        container.executeWithResource(closeable -> assertThat(closeable).isEqualTo(mockCloseable));
+
+        settableRefreshable.update(UPDATED_VALUE);
+        container.executeWithResource(closeable -> assertThat(closeable).isEqualTo(mockCloseableExtra));
+    }
+
+    @Test
+    public void previousResourceClosedAfterRefresh() throws IOException {
+        // Creating a container has the side effect of managing all resources created within it.
+        getContainer();
+
+        verifyNoInteractions(mockCloseable);
+        settableRefreshable.update(UPDATED_VALUE);
+        verify(mockCloseable).close();
+    }
+
+    @Test
+    public void latestResourceClosedAfterClose() throws IOException {
+        ReloadingCloseableContainer<Closeable> container = getContainer();
+
+        verifyNoInteractions(mockCloseable);
+        settableRefreshable.update(UPDATED_VALUE);
+        container.close();
+
+        verify(mockCloseableExtra).close();
+    }
+
+    @Test
+    public void noFurtherResourceCreatedAfterClose() {
+        ReloadingCloseableContainer<Closeable> container = getContainer();
+
+        verifyNoInteractions(mockCloseable);
+        container.close();
+        settableRefreshable.update(UPDATED_VALUE);
+
+        verify(closeableFactory, times(1)).apply(any());
```

Comment:
This is fun - this test currently fails, although the container class itself is doing the right thing(TM). Refreshables seems to be mapping twice when the map is first set up, which is odd... There's a thread in internal chat around this, where the consensus is this shouldn't be happening.